File: oolab/src/main/java/agh/ics/oop/model/MapDirection.java
package agh.ics.oop.model;

/**
 * Enum class that describes all geographical directions
 */
public enum MapDirection {
    NORTH(new Vector2d(0, 1), "Północ"),
    SOUTH(new Vector2d(0, -1), "Południe"),
    WEST(new Vector2d(-1, 0), "Zachód"),
    EAST(new Vector2d(1, 0), "Wschód");

    private final Vector2d unitVector;
    private final String directionName;

    MapDirection(Vector2d unitVector, String directionName) {
        this.unitVector = unitVector;
        this.directionName = directionName;
    }


    public String toString() {
        return directionName;
    }

    /**
     * Method to return next geographical direction clockwise
     * @return MapDirection that is next geographical direction clockwise
     */
    public MapDirection next() {
        return switch (this) {
            case NORTH -> EAST;
            case EAST -> SOUTH;
            case SOUTH -> WEST;
            case WEST -> NORTH;
        };
    }

    /**
     * Method to return next geographical direction counterclockwise
     * @return MapDirection that is next geographical direction counterclockwise
     */
    public MapDirection previous() {
        return switch (this) {
            case NORTH -> WEST;
            case WEST -> SOUTH;
            case SOUTH -> EAST;
            case EAST -> NORTH;
        };
    }

    /**
     * Method to return unit vector od direction
     * @return Vector2d object that is unit vector of direction
     */
    public Vector2d toUnitVector() {
        return unitVector;
    }
}

File: kotlin/src/main/kotlin/proj_lab2/model/MapDirection.kt
package proj_lab2.model

enum class MapDirection(private val unitVector: Vector2d, private val directionName: String) {
    NORTH(Vector2d(0, 1), "Północ"),
    SOUTH(Vector2d(0, -1), "Południe"),
    WEST(Vector2d(-1, 0), "Zachód"),
    EAST(Vector2d(1, 0), "Wschód");

    override fun toString() = directionName

    fun getUnitVector(direction: MapDirection) = direction.unitVector

    fun next(): MapDirection = when (this) {
        NORTH -> EAST
        EAST -> SOUTH
        SOUTH -> WEST
        WEST -> NORTH
    }

    fun previous(): MapDirection = when (this) {
        NORTH -> WEST
        WEST -> SOUTH
        SOUTH -> EAST
        EAST -> NORTH
    }
}